app/models/concerns/host_searching.rb

```diff
@@ -109,9 +109,8 @@ def filter_by_compliance(_filter, operator, value)
     end
 
     def filter_by_compliance_score(_filter, operator, score)
-      ids = ::Host.includes(:test_result_profiles).select do |host|
-        host.compliance_score.public_send(operator, score.to_f)
-      end
+      ids = ::Host.joins(:test_result_profiles)
+                  .where("test_results.score #{operator} ?", score.to_f)
```

Comment:
I'm not happy about the `#{operator} ?` here, but I am not sure if this can be done via Arel as the `operator` is coming from outside :confused: :shrug:

Reply:
See the related filters in this module. We can't just use the raw `operator` directly because it's not necessarily SQL. For example, users can pass `!=` or `<>` for not equal. We can limit the available operators like other scoped searches do at the top of this file.

Reply:
So what should be the allowed operators? `['>', '>=', '=', '<=', '<']`?

Reply:
Try mapping operators to Arel. There might be a method to do that already...

Definitely do not open this for an SQL Injection. Consider at least asserting the values before formatting the string.

Reply:
The `operator` method is already sanitized to the allowed operators, so that should be safe. I just want to know which ones should we allow above...

Reply:
Using the `?` format should cause ActiveRecord to sanitize the SQL, so we shouldn't need to worry about injection, but we could certainly write a test that confirms it. Any/all operators that make sense for a float comparison should be included IMO. In fact, the [default seems ok](https://github.com/wvanbergen/scoped_search/blob/master/lib/scoped_search/definition.rb#L255) if we can translate it to SQL:

```rb
['=', '>', '<', '<=', '>=', '!=', '^', '!^']  
```

Reply:
@skateman where is it sanitized? It could be much safer to do some assertion in this method to be absolutely sure.

@akofink SQL placeholders would quote it as a value, which the SQL parser does not expect for an operator (IIRC).

Reply:
Confirmed, this can be a problem in ActiveRecord:

```
[23] pry(main)> Profile.count
   (0.8ms)  SELECT COUNT(*) FROM "profiles"
=> 0
[24] pry(main)> op='=\'\' or 1=1);--'
=> "='' or 1=1);--"
[25] pry(main)> Profile.where("ref_id#{op}?", 'bar')
  Profile Load (0.6ms)  SELECT "profiles".* FROM "profiles" WHERE (ref_id='' or 1=1);--'bar')
=> [] # this would be all profiles if there were any in the DB, even those outside the user account
```

scoped_search's query language likely will not allow the operator to be injected in this way, since it explicitly validates the operator exists within the allowable operators, but it wouldn't hurt to harden (and test) this.

Reply:
The `options` argument here is NOT an SQL operator, but a string-based parsed value coming from `scoped_search` that is luckily interpretable as an SQL operator. It is not vulnerable to an SQL injection described above by @akofink because it only allows fields that are declared using the `scoped_search` method in our models. Having an operator other than the ones listed above would [raise](https://github.com/wvanbergen/scoped_search/blob/master/lib/scoped_search/definition.rb#L258) a `ScopedSearch::QueryNotSupported` exception and even the right side of the operator is sanitized using a [regular expression](https://github.com/wvanbergen/scoped_search/blob/master/lib/scoped_search/definition.rb#L261). 

If we really want to do some kind of hardening, I would imagine it like this:
```ruby
    # These are dynamically defined, so can't just pull them out from the scoped_search library :(
    NUM_OPERATORS = ['=', '>', '<', '<=', '>=', '!=', '^', '!^'].freeze

    def filter_by_compliance_score(_filter, operator, score)
      unless NUM_OPERATORS.include?(operator)
        raise ActiveRecord::StatementInvalid
      end
```

But I kinda find it pointless and also ugly :see_no_evil: 

Regarding testing, I don't think it provides any coverage to just unit-test the `filter_by_compliance_score` method with arguments that would never come from `scoped_search` anyway. I am thinking hard about higher level tests that would try to reproduce the injection that @akofink described above (or similar), but I don't really know how to formulate it in a query without making it obviously too easy for the parser to catch it. So I'm kinda lost in here what should I really do :shrug: 

Just as an example, if you do `1=1` in the search, you will get the following error without even hitting the `filter_by_compliance_score` method:
```
Field '1' not recognized for searching!
```

Reply:
As mentioned, the `operator` is scoped search one. It does not represent an SQL operator, only it resembles it. For example `^` and `!^` are not valid in numerical SQL context:
```
compliance_dev=# SELECT 1 FROM profiles where score ^ 1;
ERROR:  argument of WHERE must be type boolean, not type numeric
LINE 1: SELECT 1 FROM profiles where score ^ 1;
```

Considering that this is a public class method and the `operator` value is going directly to a raw SQL, it is our responsibility to harden and avoid any potential situation. Even if it might seem useless at the time. scoped_search might change its behavior or documentation, or one can use the method elsewhere in our app not considering the consequences. 

I highly suggest to take the `operator` value as "a switch from allowed values".

Reply:
I agree with nearly everything you said, @skateman; the risk of injection is minimal or even nonexistent. Still, what @vkrizan says about the operator not being SQL is an issue. We need some mapping from the scoped search operator to SQL.

Reply:
Yeah, it's already there